### PR TITLE
feat(secrets): wire up the import backend

### PR DESCRIPTION
**Pull request by wallyworld**
The `ImportSecrets` method on the secret service is implemented. This allows models with secrets to be migrated.
The implementation is the reverse of the previously completed method to export secrets. It takes a `SecretExport` struct with all the data and inserts into state.

## QA steps

deploy a cross model scenario with the dummy source and sink charms
create and share a charm secret and read it
create a user secret and allow a charm to read it

migrate consuming model
migrate offering model
check secrets still work
check dump-model on target controller

## Links

**Jira card:** [JUJU-5905](https://warthogs.atlassian.net/browse/JUJU-5905)

[JUJU-5905]: https://warthogs.atlassian.net/browse/JUJU-5905?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: domain/secret/service/exportimport.go [hunk 1]
```diff
@@ -140,7 +140,130 @@ func (s *SecretService) GetSecretsForExport(ctx context.Context) (*SecretExport,
...
+		}
+		if rev.ValueRef == nil {
+			if data, ok := content[rev.Revision]; ok {
+				params.Data = data
+			}
+		}
+		if i == len(revisions)-1 {
+			params.ExpireTime = md.LatestExpireTime
+		}
+		if i == 0 {
```
Comment (ycliuhw) [inline domain/secret/service/exportimport.go:210]: If it's the 1st revision, we use `CreateSecret`, other revisions use `UpdateSecret`?

File: domain/secret/service/exportimport.go [hunk 2]
```diff
@@ -140,7 +140,130 @@ func (s *SecretService) GetSecretsForExport(ctx context.Context) (*SecretExport,
...
+func (s *SecretService) importRemoteSecrets(ctx context.Context, remoteSecrets []RemoteSecret) error {
+	remoteSecretLatest := make(map[*coresecrets.URI]int)
+	for _, rs := range remoteSecrets {
+		remoteSecretLatest[rs.URI] = rs.LatestRevision
+	}
+	for uri, latest := range remoteSecretLatest {
+		if err := s.st.UpdateRemoteSecretRevision(ctx, uri, latest); err != nil {
+			return errors.Annotatef(err, "saving remote secret reference for %q", uri)
+		}
+		break
```
Comment (ycliuhw) [inline domain/secret/service/exportimport.go:258]: Why break here?

File: domain/secret/service/exportimport.go [hunk 3]
```diff
@@ -140,7 +140,130 @@ func (s *SecretService) GetSecretsForExport(ctx context.Context) (*SecretExport,
...
+	}
+	for _, rs := range remoteSecrets {
+		if err := s.st.SaveSecretConsumer(ctx, rs.URI, rs.Accessor.ID, &coresecrets.SecretConsumerMetadata{
+			Label:           rs.Label,
+			CurrentRevision: rs.CurrentRevision,
+		}); err != nil {
+			return errors.Annotate(err, "saving remote secret consumer")
```
Comment (ycliuhw) [inline domain/secret/service/exportimport.go:265]: We can include the URI and Accessor.ID in the error message.

File: domain/secret/service/exportimport.go (hunk 1 above)
Comment (wallyworld) [reply domain/secret/service/exportimport.go:210]: yup

File: domain/secret/service/exportimport.go (hunk 2 above)
Comment (wallyworld) [reply domain/secret/service/exportimport.go:258]: typo, will fix

File: domain/secret/service/exportimport.go [hunk 4]
```diff
@@ -140,7 +140,130 @@ func (s *SecretService) GetSecretsForExport(ctx context.Context) (*SecretExport,
...
+					ID:   access.Scope.ID,
+				},
+				Subject: SecretAccessor{
+					Kind: access.Subject.Kind,
+					ID:   access.Subject.ID,
+				},
+				Role: access.Role,
+			})
+			if err := s.st.GrantAccess(ctx, md.URI, p); err != nil {
+				return errors.Annotatef(err, "saving secret access for %q", md.URI.ID)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/secret/service/exportimport.go]: included subject, don't need scope

**Review by ycliuhw — APPROVED** [review]
LGTM, thanks!

**Pull request merged**